feat(providers): add NetSuite OAuth Provider

HeavenlyEntity opened the pull request:
Add NetSuite OAuth 2.0 Provider so users can use this integration with their NetSuite Account. Documentation has been added along with PNG of logo to render on Sign In. Additionally added to issue template.

<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

> _NOTE_:
>
> - It's a good idea to open an issue first to discuss potential changes.
> - Please make sure that you are _NOT_ opening a PR to fix a potential security vulnerability. Instead, please follow the [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md) to disclose the issue to us confidentially.

## ☕️ Reasoning
Add **NetSuite** OAuth provider for new and existing users looking to get started with integrating NetSuite using  **Auth.js**! Lots of people and low documentation out there to help users create such compatibility. This will be the _plug-n-play_ for those using NetSuite as there backend database provider / ERP System. 🎉 - Please let me know if I can make any adjustments!

## 🧢 Checklist

- [x] Documentation
- [x] Tests
- [x] Ready to be merged

## 🎫 Affected issues
None ✅ 

## 📌 Resources

- [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md)
- [Contributing guidelines](https://github.com/nextauthjs/.github/blob/main/CONTRIBUTING.md)
- [Code of conduct](https://github.com/nextauthjs/.github/blob/main/CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

HeavenlyEntity committed:
:sparkles: FEAT(core) Add NetSuite Provider

Add NetSuite OAuth 2.0 Provider so users can use this integration with their NetSuite Account. Documentation has been added along with PNG of logo to render on Sign In. Additionally added to issue template.

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

HeavenlyEntity commented:
When will new providers be considered to be added? This one I do have working in a production environment. Let me know if there is anything I need to do on my side or how I can help you guys to speed up the process. 🚀

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

HeavenlyEntity commented:
Hello @ThangHuuVu I see the new docs are looking great! Let me know how I can add the NetSuite doc instructions into the list of providers. Provide me with any new links to instructions on this would be great. Thank you!

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

ndom91 reviewed (commented):
Can we simplify / set as default any more of the configuration options? 

In addition, a few things regarding the `profile` optoin.

First, we'd like to always return at least the basic 4 profile fields without requiring the user to add their own callback. So the `profile` option should look somethign like this:

```ts
  profile(profile) {
    return {
      id: profile.id,
      name: profile.full_name,
      email: profile.email,
      image: profile.avatar_url,
    }
  },
```

In addition, it looks like the `token` argument to the callback isn't used in your example etiher, can we just drop that?

Finally, I see a lot of formatting / whitespace issues in the code at the bottom, and the JSDoc comment should follow the normal JSDoc formatting. See the `src/providers/netlify.ts` (or any other provider) as an example :pray:

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

ndom91 commented:
> Hello @ndom91 👋🏼 ! I have went ahead and removed the token arg from the example and reformatted the documentation. Unfortunately with NetSuite they do not supply an out-of-the-box callback for us so users would have to create their own RESTlet callback utilizing `N/runtime`. I can provide an example if needed. Looking forward to your reply!

Yeah an example would be great then, thanks!

ndom91 commented:
Also it seems like something got out of whack here. Maybe your source branch just needs to be updated with `main`, but oddly Github isn't offering the 1-click "update to 'main'" option, you'll have to do it manually

HeavenlyEntity commented:
@ndom91 I have updated the branch and provided an example of the RESTlet that needs deployed in the NetSuite account to use as a callback. I have specified additional instructions for NetSuite users who are not familiar with the deployment process. Let me know if there are any additional adjustments needed. 🤝

ndom91 commented:
@HeavenlyEntity hmm there's still something wrong. The changes to `apps/dev/sveltekit/src/auth.ts` and `.github/workflows/release.yml` and the issue_templste shouldn't be included here, for example. 

Those were updates we made and merged into main a while ago. That's why i was thinking that your source branch is just out of date, however normally github offers a 1-click update source branch button which isn't the case here atm

HeavenlyEntity committed:
:recycle: Update(provider) NetSuite

Provide more information with an example of a RESTlet and better formatting

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

HeavenlyEntity commented:
@ndom91 Branch is fixed and up-to-date! ⬆️

ndom91 reviewed (approved):
LGTM! Thanks for your patience!

ndom91 commented on packages/core/src/providers/netsuite.ts:
```diff
@@ -0,0 +1,238 @@
+      }
+    },
+    userinfo: `${config.userInfoUrl}`,
+    profile: (profile: Object, token) => config.profileCallback(profile, token),
```
Ah sorry, I thought we'd changed this already. We shouldn't require the user to pass a profile callback. Lets use the default here

```suggestion
-    profile: (profile: Object, token) => config.profileCallback(profile, token),
+    profile(profile) {
+      return {
+        id: profile.id.toString(),
+        name: profile.name ?? profile.username,
+        email: profile.email,
+        image: profile.avatar_url,
+      }
+    },
```

And then folks can always still use the `profile` option of the provider to override anything if they want to add more fields, etc.

```ts
  providers: [
    NetSuite({
      profile(profile) {
        return { role: profile.role ?? "user", ... }
      },
    })
  ],
```

HeavenlyEntity replied:
I have implemented new defaults @ndom91 along with new documentation. Let me know if this is in anyway confusing. Lol I have also changed some props on the options to mimic the Salesforce provider.

ndom91 commented on packages/core/src/providers/netsuite.ts:
```diff
@@ -0,0 +1,238 @@
+ *
+ * :::
+ */
+export default function NetSuite(
+  config: OAuthNetSuiteOptions
+): string | object {
```
Also this should return an `OAuthConfig` and take a `OAuthUserConfig` type object. You can import these from `.../providers/index.js`:

```ts
import type { OAuthConfig, OAuthUserConfig } from "./index.js"
```

You'll have to define `NetSuiteProfile` above somewhere, this shuold cover everythign NetSuite cuold return and a user might want to put on their profile. See `providers/gitlab.ts` for example.



```suggestion
export default function NetSuite<P extends NetSuiteProfile>(
  options: OAuthUserConfig<P>
): OAuthConfig<P> {
```

HeavenlyEntity commented:
> Sorry, shuoldn't have approved it so quickly, there are a few bigger things we need to clean up..
> 
> Also would you mind if I push a commit to your branch myself? Just cleaning up some of the text copy / wording?

That is fine @ndom91 😊. I would encourage you to reach out if you have any questions regarding NetSuite. I can address those over email. So we're not clogging up the PR conversation 😂. 

Looking at your review comments I will take a look at those later this evening 🦾. Thank you for the feedback.

HeavenlyEntity committed:
Update packages/core/src/providers/netsuite.ts

Co-authored-by: Nico Domino <yo@ndo.dev>

HeavenlyEntity committed:
:recycle: Reactor(provider) NetSuite `profile` with defaults

Use the default `runtime` schema from NetSuite. Then users can add more later on.

HeavenlyEntity committed:
Merge branch 'feat-add-netsuite-oauth' of https://github.com/HeavenlyEntity/next-auth into feat-add-netsuite-oauth

HeavenlyEntity committed:
:recycle: REF(provider) Update types and docs

HeavenlyEntity committed:
:memo: Update(provider) NetSuite Docs

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

ndom91 commented on packages/core/src/providers/netsuite.ts:
```diff
@@ -0,0 +1,248 @@
+        grant_type: "authorization_code",
+      }
+    },
+    userinfo: `${userInfo}`,
```
Is there a reason you're putting this string in a template literal? 

We coudl avoid renaming it to `userInfo` and just have this line be `userinfo`

```suggestion
    userinfo,
```

HeavenlyEntity replied:
Yes @ndom91 it's more so preference for readability. The template literal can be removed. There is an alias at the top where the destructuring is placed `userinfo: userInfo`. Having all lowercase for just that key and the rest aren't. It keeps the consistency and if it later gets changed from a lowercase key to a camelcase one it will be easy to refactor ensuring backwards compatibility.

ndom91 replied:
Right my point was just, if we skip the alias all together we can simplify this a bit (even though its then no longer camelCase haha)

ndom91 replied:
Either way, not a big deal. Let's try to get this in. 

Is there a public site i can make an account at that's using netsuite to test this?

HeavenlyEntity replied:
Haha I totally understand that 😄. There isn't a public site to test no. NetSuite is all blocked behind a paywall. But if you have the repo set up I can give you a test set of data? I do also have a working version of this used in production on a nextjs project. I can send you that as well? Whichever path works best for you. Lmk 😎

HeavenlyEntity replied:
@ndom91 let me know if you need assistance. I can test this with you if need be. Let me know!
![IGotYouThumbsUpGIF](https://github.com/nextauthjs/next-auth/assets/39228141/b62c2c79-e3e9-4ece-ad5c-4473687f7ae6)

ndom91 replied:
Hmm okay thats what I feared, that it was very enterprisey and not available to test for us.

Okay not a huge deal, we can merge it and deal with things as they come in. I assume NetSuite isn't going to be a hugely popular oauth provider :sweat_smile:

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

HeavenlyEntity committed:
:zap: Optimize(provider) NetSuite shorthand defaults

Optimize config option defaults and reading simplicity by object shorthand sets to the config object. Creates easier dynamic for future adjustments as NetSuite evolves.

HeavenlyEntity committed:
:pencil: Update(provider) comments for Sandboxes

Update issuer comment for the inclusion of Sandboxes. As the account ID for sandboxes default to `1234567_SB(1 or 2)` when in the url it needs parsed as a `-`. So this has been included for users using sandbox accounts to test.

HeavenlyEntity committed:
:wrench: Update(provider) `N/runtime` profile type

Update the NetSuite profile return type with all the new properties a part of the new 2024.3 release.

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

ndom91 commented on packages/core/src/providers/netsuite.ts:
```diff
@@ -0,0 +1,252 @@
+ *
+ * :::
+ */
+export default function NetSuite<P extends NetSuiteProfile>(
```
One other tiny tweak, if we name this `options`, you can pull out whatever you need on like `213`, and then just pass the rest of the `options` in under hte `options` key on like `250` and don't need to pull out `clientId` and `clientSecret` additionally, for example. 

i.e.


```suggestion
export default function NetSuite<P extends NetSuiteProfile>(
  config: OAuthNetSuiteOptions<P>
): OAuthConfig<P> {
  const { issuer, userinfo: userInfo, prompt = 'none', scope = 'restlets rest_webservices', clientId, clientSecret } = options

  return {
    id: "netsuite",
    name: "NetSuite",
    type: "oauth",
    checks: ["state"],
    authorization: {
      url: `https://${issuer}.app.netsuite.com/app/login/oauth2/authorize.nl`,
      params: {
        client_id: clientId,
        prompt: prompt,
        response_type: 'code',
        scope,
      }
    },
    token: {
      url: `https://${issuer}.suitetalk.api.netsuite.com/services/rest/auth/oauth2/v1/token`,
      params: {
        grant_type: "authorization_code",
      }
    },
    userinfo: userInfo,
    profile(profile) {
      // This is the default runtime.getCurrentUser() object returned from the RESTlet or SUITELet
      return {
        id: profile.id,
        name: profile.name,
        email: profile.email,
        location: profile.location,
        role: profile.role,
        contact: profile?.contact
      }
    },
    style: { logo: "/netsuite.png", bg: "#3a4f5f", text: "#fff" },
    options
  }
}
```

HeavenlyEntity replied:
That is fine. You can commit this suggestion if you'd like. 👍🏼  @ndom91

ndom91 replied:
@HeavenlyEntity I pushed that fix and cleaned up a few other small thigns :blush:

Can you give it another test to make sure it still works? If it does we can merge it finally haha :tada:

HeavenlyEntity replied:
@ndom91 Thank you for getting this in! I noticed the color is not right lol. Can you update that to be this color for the background _#3a4f5f_ ➡️ _**#18191f**_ ? Hope that can be committed directly. Thank you @ndom91! Wanna get it perfect before the release! 🚀

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

ndom91 committed:
Merge branch 'main' into feat-add-netsuite-oauth

ndom91 committed:
fix: cleanup Netsuite provider

ndom91 committed:
fix: add svg netsuite logo

ndom91 committed:
feat: add Netsuite.mdx new docs page

ndom91 committed:
Merge branch 'main' into feat-add-netsuite-oauth

HeavenlyEntity committed:
Merge branch 'main' into feat-add-netsuite-oauth

ndom91 committed:
Merge branch 'main' into feat-add-netsuite-oauth

ndom91 commented:
Thanks for your patience with this! It'll be available in the next releases we cut (i.e. `next-auth@5.0.0-beta.18`, `@auth/sveltekit@1.0.2`, etc.), unfortunately I can't say exactly when that'll be but my best guess is within a week or two max :pray:

HeavenlyEntity commented:
Hi @mbaquerizo ! Yeah I have not had luck with the OIDC method because NetSuite is really strict with permissions and roles. You have to set up up the permissions to allow users to auth this way. Which is extremely tedious. My approach utilizes OAuth 2.0 which NetSuite recommends. What is cool about this method the token returned from the `profile` method allows you to auth each restlet as the signed in user in the `api` routes using the `getServerSession`. Super cool! 🚀

HeavenlyEntity commented:
> Thanks for your patience with this! It'll be available in the next releases we cut (i.e. `next-auth@5.0.0-beta.18`, `@auth/sveltekit@1.0.2`, etc.), unfortunately I can't say exactly when that'll be but my best guess is within a week or two max 🙏

@ndom91 Thank you for getting this in! I noticed the color is not right lol. Can you update that to be this color for the background _#3a4f5f_ ➡️ _**#181a1b**_ ? Then the text to **#fbfbfb**. Hope that can be committed directly. Thank you @ndom91! Wanna get it perfect before the release! 🚀
